grid-client/deployer/deployment_deployer.go

```diff
@@ -185,24 +208,189 @@ func (d *DeploymentDeployer) updateStateFromDeployments(ctx context.Context, dl
 		}
 
 		if dlData.Name == dl.Name {
-			dl.NodeDeploymentID[dl.NodeID] = newDl.ContractID
+			if newDl.ContractID != 0 {
+				dl.NodeDeploymentID[dl.NodeID] = newDl.ContractID
+				dl.ContractID = newDl.ContractID
+				d.tfPluginClient.State.StoreContractIDs(dl.NodeID, []uint64{dl.ContractID})
+			}
 		}
 	}
 
-	if contractID, ok := dl.NodeDeploymentID[dl.NodeID]; ok && contractID != 0 {
-		dl.ContractID = contractID
-		if !workloads.Contains(d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID], dl.ContractID) {
-			d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID] = append(d.tfPluginClient.State.CurrentNodeDeployments[dl.NodeID], dl.ContractID)
+	return nil
+}
+
+func (d *DeploymentDeployer) calculateNetworksUsedIPs(ctx context.Context, dls []*workloads.Deployment) (map[string]map[uint32][]byte, error) {
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	var errs error
+	usedHosts := make(map[string]map[uint32][]byte)
+
+	// calculate used host IDs per network
+	for _, dl := range dls {
+		mu.Lock()
+		if _, ok := usedHosts[dl.NetworkName]; !ok {
+			usedHosts[dl.NetworkName] = make(map[uint32][]byte)
+		}
+
+		if _, ok := usedHosts[dl.NetworkName][dl.NodeID]; ok {
+			continue
+		}
+		mu.Unlock()
+
+		wg.Add(1)
+		go func(dl *workloads.Deployment) {
+			defer wg.Done()
+
+			nodeClient, err := d.tfPluginClient.NcPool.GetNodeClient(d.tfPluginClient.SubstrateConn, dl.NodeID)
+			if err != nil {
+				mu.Lock()
+				defer mu.Unlock()
+				errs = multierror.Append(errs, errors.Wrapf(err, "could not get node client for node %d", dl.NodeID))
+				return
+			}
+
+			privateIPs, err := nodeClient.NetworkListPrivateIPs(ctx, dl.NetworkName)
+			if err != nil {
+				mu.Lock()
+				defer mu.Unlock()
+				errs = multierror.Append(errs, errors.Wrapf(err, "could not list private ips from node %d", dl.NodeID))
+				return
+			}
+
+			network := d.tfPluginClient.State.Networks.GetNetwork(dl.NetworkName)
+			ipRange := network.GetNodeSubnet(dl.NodeID)
+
+			_, ipRangeCIDR, err := net.ParseCIDR(ipRange)
+			if err != nil {
+				mu.Lock()
+				defer mu.Unlock()
+				errs = multierror.Append(errs, errors.Wrapf(err, "invalid ip range %s", ipRange))
+				return
+			}
+
+			// get used host IDs of the deployment node using the private IPs list of the node
+			for _, privateIP := range privateIPs {
+				parsedPrivateIP := net.ParseIP(privateIP).To4()
+
+				if ipRangeCIDR.Contains(parsedPrivateIP) {
+					mu.Lock()
+					defer mu.Unlock()
+					usedHosts[dl.NetworkName][dl.NodeID] = append(usedHosts[dl.NetworkName][dl.NodeID], parsedPrivateIP[3])
```

Comment:
we can move all that to a new function that return error and used hosts so we can lock once instead of locking appending unlocking returning multiple times